File: cursory/Cursory.java
package cursory;

import com.sun.jna.Platform;

public abstract class Cursory implements AutoCloseable {

    public class Event {
        public String eventType;
        public String which;
        public Event(String eventType, String which) {
            this.eventType = eventType;
            this.which = which;
        }
        public Event() { this("", ""); }
    }

    public class XY {
        public int x;
        public int y;
        public XY(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    abstract public XY getSize() throws Exception;
    abstract public void restoreMode() throws Exception;
    abstract public void enableRawMode() throws Exception;
    abstract public void render(Iterable<RenderAction> actions);
    public void close() throws Exception { restoreMode(); }

    public static boolean isTerminal(int fileDescriptor) throws Exception {
        if (Platform.isMac()) {
            return CursoryUnix.isatty(fileDescriptor);
        } else {
            throw new Exception("Unsupported platform");
        }
    }

    public static Cursory getTerminal(int fileDescriptor) throws Exception {
        if (Platform.isMac()) {
            return new CursoryUnix(fileDescriptor);
        } else {
            throw new Exception("Unsupported platform");
        }
    }
}

File: clojure/cursory.clj
(ns cursory
  (:require [clojure.core.match :refer [match]])
  (:import cursory.Cursory cursory.RenderAction))

(defn get-terminal []
  (Cursory/getTerminal 0))

(defn get-size [^Cursory cursory]
  (let [size (.getSize cursory)]
    (when size [(.x size) (.y size)])))

(defn get-cursor-pos [^Cursory cursory]
  (let [pos (.getCursorPos cursory)]
    (when pos [(.x pos) (.y pos)])))

(defmacro with-raw-mode [cursory & body]
  `(let [cursory# ~cursory]
     (try (do (.enableRawMode cursory#)
              ~@body)
          (finally (.restoreMode cursory#)))))

(defn read-event [^Cursory cursory]
  (let [event (.readEvent cursory)
        which (.which event)]
    (case (.eventType event)
      "" "(none)"
      "rune" which
      "specialkey" which
      "mousebutton" "mousebutton"
      "resize" "resize")))

(defn render! [^Cursory cursory actions]
  (. cursory render
     (map (fn [action]
            (match action
              [:clear-to-line-end]
              (RenderAction/clearToLineEnd)
              [:clear-to-screen-end]
              (RenderAction/clearToScreenEnd)
              [:go-abs x y]
              (RenderAction/goAbs ^int x ^int y)
              [:text s]
              (RenderAction/text ^String s)))
          actions)))
